cmd/kubeadm/app/cmd/phases/join/kubelet.go

```diff
@@ -93,13 +93,22 @@ func getKubeletStartJoinData(c workflow.RunData) (*kubeadmapi.JoinConfiguration,
 // runKubeletStartJoinPhase executes the kubelet TLS bootstrap process.
 // This process is executed by the kubelet and completes with the node joining the cluster
 // with a dedicates set of credentials as required by the node authorizer
-func runKubeletStartJoinPhase(c workflow.RunData) error {
+func runKubeletStartJoinPhase(c workflow.RunData) (err error) {
 	cfg, initCfg, tlsBootstrapCfg, err := getKubeletStartJoinData(c)
 	if err != nil {
 		return err
 	}
 	bootstrapKubeConfigFile := kubeadmconstants.GetBootstrapKubeletKubeConfigPath()
 
+	defer func() {
+		// Deletes the bootstrapKubeConfigFile, so the credential used for TLS bootstrap are removed from disk
+		deleteErr := os.Remove(bootstrapKubeConfigFile)
+		// Only return a delete error if we weren't already returning something else
+		if err == nil && deleteErr != nil {
+			err = errors.Wrapf(deleteErr, "error deleting %s", bootstrapKubeConfigFile)
+		}
+	}()
```

Comment:
I agree with deferring the delete before the write call.
If WriteToDisk returns an error, the delete will still be called as a best-effort action and will not clobber the existing error message.